plugins/source_scheduler_dfs.go

```diff
@@ -0,0 +1,192 @@
+package plugins
+
+import (
+	"context"
+	"fmt"
+	"runtime/debug"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/cloudquery/plugin-sdk/caser"
+	"github.com/cloudquery/plugin-sdk/helpers"
+	"github.com/cloudquery/plugin-sdk/schema"
+	"github.com/cloudquery/plugin-sdk/specs"
+	"github.com/getsentry/sentry-go"
+	"github.com/thoas/go-funk"
+	"golang.org/x/sync/semaphore"
+)
+
+const (
+	minTableConcurrency    = 1
+	minResourceConcurrency = 100
+)
+
+func (p *SourcePlugin) syncDfs(ctx context.Context, spec specs.Source, client schema.ClientMeta, tables schema.Tables, resolvedResources chan<- *schema.Resource) {
+	// current DFS supports only parallization for top level tables and resources.
+	// it is possible to extend support for multiple levels but this require benchmarking to find a good fit on how to split
+	// gorourtines for each level efficiently.
+	// This is very similar to the concurrent web crawler problem with some minor changes.
+	// We are using DFS to make sure memory usage is capped at O(h) where h is the height of the tree/subchilds.
+	tableConcurrency := spec.Concurrency / minResourceConcurrency
+	if tableConcurrency < minTableConcurrency {
+		tableConcurrency = minTableConcurrency
+	}
+	resourceConcurrency := tableConcurrency * minResourceConcurrency
+
+	p.tableSem = semaphore.NewWeighted(int64(tableConcurrency))
+	p.resourceSem = semaphore.NewWeighted(int64(resourceConcurrency))
```

Comment:
We might want to create a resource semaphore per table instead of sharing it across tables, as this means one branch of the DFS tree can block other branches (e.g. table 1 can block table 2 if it has many resources to handle in the current batch).

Reply:
this wont block because unlike `table` semaphores, `resource` semaphore are being released the moment resource is resolved as we use a bit more memory for that (still capped).

Reply:
Yes, agree it won't block, just might impact concurrency as table 2 won't finish resolving before table 1 is finished.